lite2/client.go

```diff
@@ -0,0 +1,374 @@
+// and calls VerifyHeader.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeaderAtHeight(height int64, now time.Time) error {
+	if c.trustedHeader.Height >= height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", height, c.trustedHeader.Height)
+	}
+
+	// Request the header and the vals.
+	newHeader, newVals, err := c.fetchHeaderAndValsAtHeight(height)
+	if err != nil {
+		return err
+	}
+
+	return c.VerifyHeader(newHeader, newVals, now)
+}
+
+// VerifyHeader verifies new header against the trusted state.
+//
+// SequentialVerification: verifies that 2/3 of the trusted validator set has
+// signed the new header. If the headers are not adjacent, **all** intermediate
+// headers will be requested.
+//
+// SkippingVerification(trustLevel): verifies that {trustLevel} of the trusted
+// validator set has signed the new header. If it's not the case and the
+// headers are not adjacent, bisection is performed and necessary (not all)
+// intermediate headers will be requested. See the specification for the
+// algorithm.
+//
+// If the trusted header is more recent than one here, an error is returned.
+func (c *Client) VerifyHeader(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	if c.trustedHeader.Height >= newHeader.Height {
+		return errors.Errorf("height #%d is already trusted (last: #%d)", newHeader.Height, c.trustedHeader.Height)
+	}
+
+	var err error
+	switch c.mode {
+	case sequential:
+		err = c.sequence(newHeader, newVals, now)
+	case skipping:
+		err = c.bisection(c.trustedHeader, c.trustedNextVals, newHeader, newVals, now)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Update trusted header and vals.
+	nextVals, err := c.primary.ValidatorSet(newHeader.Height + 1)
+	if err != nil {
+		return err
+	}
+	return c.updateTrustedHeaderAndVals(newHeader, nextVals)
+}
+
+func (c *Client) sequence(newHeader *types.SignedHeader, newVals *types.ValidatorSet, now time.Time) error {
+	// 1) Verify any intermediate headers.
+	var (
+		interimHeader *types.SignedHeader
+		nextVals      *types.ValidatorSet
+		err           error
+	)
+	for height := c.trustedHeader.Height + 1; height < newHeader.Height; height++ {
+		interimHeader, err = c.primary.SignedHeader(height)
+		if err != nil {
+			return errors.Wrapf(err, "failed to obtain the header #%d", height)
+		}
+
+		err = Verify(c.chainID, c.trustedHeader, c.trustedNextVals, interimHeader, c.trustedNextVals, c.trustingPeriod, now, c.trustLevel)
+		if err != nil {
+			return errors.Wrapf(err, "failed to verify the header #%d", height)
+		}
+
+		// Update trusted header and vals.
+		if height == newHeader.Height-1 {
+			nextVals = newVals
+		} else {
+			nextVals, err = c.primary.ValidatorSet(height + 1)
+			if err != nil {
+				return errors.Wrapf(err, "failed to obtain the vals #%d", height+1)
+			}
+		}
+		if !bytes.Equal(interimHeader.NextValidatorsHash, nextVals.Hash()) {
```

Comment:
... see here we verify the next validator set hash explicitly, and then call `updateTrustedHeaderAndVals`.